pkg/cloud/services/ec2/instances.go

```diff
@@ -259,6 +222,82 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte) (*i
 	return out, nil
 }
 
+// findSubnet attempts to retrieve a subnet ID in the following order:
+// - subnetID specified in machine configuration,
+// - subnet based on filters in machine configuration
+// - subnet based on the availability zone specified,
+// - default to the first private subnet available.
+func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
+	// Prefer AWSMachine.Spec.FailureDomain for now while migrating to the use of
+	// Machine.Spec.FailureDomain. The MachineController will handle migrating the value for us.
```

Comment:
This should likely be reversed, otherwise we could potentially ignore the failure domain spreading logic in CAPI.

Reply:
That bit was copied over from the existing code: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/services/ec2/instances.go#L162-L167

I'll do some digging to see if it needs updating

Reply:
I think this is working as expected. 
- In most cases, the failureDomain on the AWSMachine will be nil and the failure domain spreading logic will work as expected. 
- If a failureDomain is defined by the user in the AWSMachineTemplate then all machines would be created in that AZ regardless of the rotation that CAPI attempts.

Reply:
> If a failureDomain is defined by the user in the AWSMachineTemplate then all machines would be created in that AZ regardless of the rotation that CAPI attempts.

This is the situation I'm most worried about, consider someone has a v1alpha2-based cluster and upgrades to v1alpha3, when they perform the steps for adopting their existing control plane components into a KubeadmControlPlane they accidentally set AWSMachineTemplate.Spec.Template.Spec.FailureDomain to a specific failure domain, then any additional actions taken on their control plane will now result in a control plane that is entirely stored in a single failure domain overriding the spreading logic in KubeadmControlPlane.

Reply:
Ah I didn't think about the upgrade flow and users accidentally restricting the failure domain. Flipping the logic would mean the user can't dictate the failure domain for control plane with that field. But they still have ways through the subnet field so that sounds reasonable. Will update.